### DISCUSSION TITLE
[Potential Bug] Structure behavior

### DISCUSSION THREAD

**Discussion by IchiruTake** [rdkit/rdkit#4937] [Q&A] [answered]
The structure represented here is `On1c(-c2ccccc2)c(-c2c(-c3ccccc3)[nH]c3ccccc23)c2ccccc21` which is 3-(2-Phenyl-1H-indole-3-yl)-2-phenyl-1H-indole-1-ol, according to PubChem (Link below). When I convert this structure, I have found that the mol cannot be sanitized: `rdkit.Chem.rdchem.KekulizeException: Can't kekulize mol.  Unkekulized atoms: 10 11 18 19 20 21 22 23 24`

When I tested on ChemDraw, it also shown the similar result, while I believe the current structure is fine. To be more specific, the error in ChemDraw (parse RDKit's SMILES) is that `there is a charge or valence error somewhere in this aromatic system`. The highlighted system is the 5-membered ring that contained `N-O` bond. 

Basic SMILES: `C1=CC=C(C=C1)C2=C(C3=CC=CC=C3N2)C4=C(N(C5=CC=CC=C54)O)C6=CC=CC=C6`
Canonical SMILES by RDKit (Converted by `Basic SMILES`): `On1c(-c2ccccc2)c(-c2c(-c3ccccc3)[nH]c3ccccc23)c2ccccc21`

Link: https://pubchem.ncbi.nlm.nih.gov/compound/10894652#section=Structures
Windows 10 Pro - Python 3.8.10 - RDKit: 2021.03.03 & 2021.09.03

* Remark: Two different RDKit's results in same error as above
@greglandrum

**Comment by greglandrum** [comment, marked as answer]
I can't reproduce this. Can you please share exactly the command you execute which causes the failure?

**Reply by IchiruTake** [reply]
``` python
from typing import List

from rdkit.Chem import AddHs, MolFromSmiles, MolToSmiles, SanitizeMol
from rdkit.Chem.rdchem import RWMol

smi = 'C1=CC=C(C=C1)C2=C(C3=CC=CC=C3N2)C4=C(N(C5=CC=CC=C54)O)C6=CC=CC=C6'
canon_smiles = MolToSmiles(MolFromSmiles(smi))
bond_index = 36  # Bond O-H: Oxygen directly connects to Nitrogen

mol = RWMol(AddHs(MolFromSmiles(canon_smiles)))
oneDegreeAtomFilter: List[int] = []
for bond in mol.GetBonds():
    if bond.GetIdx() not in (bond_index,):
        if bond.GetBeginAtom().GetSymbol() == 'H':
            oneDegreeAtomFilter.append(bond.GetBeginAtom().GetIdx())
        if bond.GetEndAtom().GetSymbol() == 'H':
            oneDegreeAtomFilter.append(bond.GetEndAtom().GetIdx())

oneDegreeAtomFilter.sort(reverse=True)
print(oneDegreeAtomFilter)
for atomIndex in oneDegreeAtomFilter:
    mol.RemoveAtom(atomIndex)

# UnkekulizedAtoms = [10, 11, 18, 19, 20, 21, 22, 23, 24]
# for AtomIdx in UnkekulizedAtoms:
#     print(mol.GetAtomWithIdx(AtomIdx).GetSymbol())
SanitizeMol(mol)
```

**Reply by IchiruTake** [reply]
> I can't reproduce this. Can you please share exactly the command you execute which causes the failure?

The code above I have used for various structure in Comprehensive Handbook of BDE by Yu-Ran Luo (2007), but this is the first time I have observed this type of error in this structure.

If you have ChemDraw, you can copy this SMILES: `On1c(-c2ccccc2)c(-c2c(-c3ccccc3)[nH]c3ccccc23)c2ccccc21` and used the function: `CheckStructure` to see this error. This error will raise even if I did not make any touch on atom-removal function in ChemDraw (just parse the SMILES and validate the structure). Meanwhile, doing the same with PubChem's SMILES is perfectly fine (no error arisen).

**Reply by IchiruTake** [reply]
Note: The following error can still be arised even I paste the PubChem's basic SMILES. The bond index is 50, which correspond to atom 24 and 45

**Reply by greglandrum** [reply]
If you modify the molecule then you are also responsible for making sure that the modified molecule is correct.
You need to be very careful when removing bonds from aromatic heteroatoms, because this can lead to the type of kekulization errors you are seeing above.

It looks like the goal of your code is to just add hydrogens to the O atom (atom 24) in your molecule. This is much more easily done like this:
```
In [17]: m = Chem.MolFromSmiles('C1=CC=C(C=C1)C2=C(C3=CC=CC=C3N2)C4=C(N(C5=CC=CC=C54)O)C6=CC=CC=C6')

In [18]: mh = Chem.AddHs(m,onlyOnAtoms=[24])

In [19]: Chem.SanitizeMol(mh)
Out[19]: rdkit.Chem.rdmolops.SanitizeFlags.SANITIZE_NONE

In [20]: Chem.MolToSmiles(mh)
Out[20]: '[H]On1c(-c2ccccc2)c(-c2c(-c3ccccc3)[nH]c3ccccc23)c2ccccc21'
```

**Comment by cottonkiet** [comment]
Hi all,
I tried to reproduce your result since I have similar problems with a chemical ‘COc1ccc2nc(C)c(C=O)c2c1’
I am also seeing the error of 'Can't kekulize mol.  Unkekulized atoms: 0 1 2 3 4 6 7 8 10'.
However, this structure is perfectly fine with ChemDraw or ChemAxon. What is the problem with this structure then?
I am getting this structure because it comes from a reaction below:

```
from rdkit import Chem
from rdkit.Chem import AllChem
react = Chem.MolFromSmiles('COc1ccc2n(C(=O)c3ccc(Cl)cc3)c(C)c(C=O)c2c1') 
display(react)
scheme = '[#7;$(*([#6]=O)-,:c):3]-[#6;R0:1](-[#6:4])=[O:2]>>[#6:4]-[#6:1](-[#8])=[O:2].[*:3]'
rxn = AllChem.ReactionFromSmarts(scheme)
display(rxn)
prod = rxn.RunReactants([react])
len(prod)
display(prod[0][0])
```

Does anyone have an idea how to fix the problem? It works fine with other reactant such as 
`react = Chem.MolFromSmiles('CCc1cccc(C)c1N(C(C)COC)C(=O)CCl')`

**Reply by IchiruTake** [reply]
Actually the following code I did was to assign special attribute onto the bond I want, including the hydrogen-based bond. and remove all unused hydrogen bond. After `SanitizeMol`, I replace the condition by this:

```python
                    if beginAtom.GetSymbol() == 'H':
                        if not (endAtom.GetIsAromatic() and endAtom.GetSymbol() != 'C'):
                            removedAtomAndBond.append(beginAtom.GetIdx())
                    if endAtom.GetSymbol() == 'H':
                        if not (beginAtom.GetIsAromatic() and beginAtom.GetSymbol() != 'C'):
                            removedAtomAndBond.append(endAtom.GetIdx())

```

**Reply by wopozka** [reply]
@cottonkiet 
Your SMILES (COc1ccc2nc(C)c(C=O)c2c1) is actually wrong, you are missing hydrogen at aromatic nitrogen atom. Without hydrogen rdkit is not able to create Kekule form. Marvin deals with it just by not adding hydrogen to nitrogen, rdkit is more strict. Yours SMILES should look like COc1ccc2[nH]c(C)c(C=O)c2c1.
To make it working with the reaction you should modify your reaction SMARTS, to force formation of nH product on RHS (something like below should work):
scheme = '[#7;$(*([#6]=O)-,:c):3]-[#6;R0:1](-[#6:4])=[O:2]>>[#6:4]-[#6:1](-[#8])=[O:2].[*H:3]'

**Reply by cottonkiet** [reply]
Dear wopozka,
Thanks for your answer. 
`COc1ccc2[nH]c(C)c(C=O)c2c1` is the more correct smiles for the product. What do you mean by "Marvin deals with it just by not adding hydrogen to nitrogen"? it looks to me that Marvin automatically added a H to n... in the original smiles (COc1ccc2nc(C)c(C=O)c2c1)
The scheme you provided forced the implicit H to be 1 on the product. it works fine with secondary amines, but with primary amine below, it fail. 
`CCc1cccc(C)c1NC(=O)CCl`

**Reply by wopozka** [reply]
@cottonkiet 
As I said, Marvin presents structure in aromatic, not kekule form, therefore presence of hydrogen is not necessary. To calculate kekule form you need information about presence of hydrogen. In case of pyrole ring it can be guessed, but if there are more nitrogens then it might be tricky. RDkit does not want to guess, therefore kekulization error.  You are right about Marvin, it removes hydrogen even if it is present in smiles. When you paste smiles with hydrogen, the hydrogen is present in the structure until you press mouse button and then it is removed.
Regarding your reaction, it is difficult to deal with your problem using only one reaction smarts. You definitely need separate one for aromatic amines, and aliphatic amines. For aromatic you need to define hydrogens on RHS, for aliphatic it is probably not necessary, but I recommend checking. In worst case scenario you need separate smiles for hydrolysis to secondary and primary amines.  I would go for something like (I removed recursive SMARTS defs, as I am not fully sure what you want to achieve)
[nX3H0:3]-[CX3R0:1](-[#6:4])=[OX1:2]>>[*:4]-[C:1](-[OH])=[O:2].[nH:3]
[NX3H0,NX3H1:3]-[CX3R0:1](-[#6:4])=[OX1:2]>>[*:4]-[C:1](-[OH])=[O:2].[N:3]

**Reply by cottonkiet** [reply]
Thanks so much for your detailed expression, and two reaction schemes works!
